### PR TITLE
FS2: StatsD reporter for metrics

### PR DESCRIPTION
**Pull request by istreeter**
This adds a custom dropwizard reporter -- which actually has more functionality that what is needed for Enrich! (e.g. histograms, timers).  But we could re-use the reporter for other apps, where we may need the extra functionality.  The new reporter is backed by the [dogstatsd client](https://github.com/DataDog/java-dogstatsd-client), which is non-blocking and runs on a dedicated thread.

There are some features I considered adding, but chose not to yet:
- Option to have both statsd and stdout reporters both running at the same time.
- Option to filter out some metrics from the reporter
- Option to configure the time unit of the metrics (seconds vs millis)

I also chose *not* to generate a random UUID to be a `worker_id` tag on the metrics.  I believe it is possible to add worker identifiers through configuration at run time.  For example, in a Kubernetes environment we can add the pod_name, which uniquely identifies the worker:
```
tags = {
  pod_name = ${HOSTNAME}
}
```

### CODE REVIEW

File: modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/io/Metrics.scala [hunk 1]
```diff
@@ -47,10 +49,10 @@ trait Metrics[F[_]] {
...
 
   val LoggerName = "enrich.metrics"
-  val LatencyGaugeName = "enrich.metrics.latency"
-  val RawCounterName = "enrich.metrics.raw.count"
-  val GoodCounterName = "enrich.metrics.good.count"
-  val BadCounterName = "enrich.metrics.bad.count"
+  val LatencyGaugeName = "snowplow.enrich.metrics.latency"
+  val RawCounterName = "snowplow.enrich.metrics.raw"
+  val GoodCounterName = "snowplow.enrich.metrics.good"
+  val BadCounterName = "snowplow.enrich.metrics.bad"
```
Comment (istreeter) [inline modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/io/Metrics.scala:55]: I am slightly worried it is a breaking change to rename these metrics.  On the other hand, fs2 is currently only released as "public beta", and until now metrics have been used only for debugging.  I think the new names are more appropriate when we start sending metrics to a statsd server.

File: modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/io/StatsDReporter.scala [hunk 2]
```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2020 Snowplow Analytics Ltd. All rights reserved.
```
Comment (chuwy) [inline modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/io/StatsDReporter.scala:2]: 2021

File: modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/config/io.scala [hunk 3]
```diff
@@ -98,4 +100,26 @@ object io {
...
+      hostname: String,
+      port: Int,
+      tags: Map[String, String]
+    ) extends MetricsReporter
+
+    implicit val finiteDurationEncoder: Encoder[FiniteDuration] =
+      implicitly[Encoder[String]].contramap(_.toString)
+
+    implicit val metricsReporterDecoder: Decoder[MetricsReporter] =
+      deriveConfiguredDecoder[MetricsReporter]
```
Comment (chuwy) [inline modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/config/io.scala:121]: I thought we were going to add a fallback option with `metricsReportPeriod` to make configuration foward-compatible?

File: modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/config/io.scala [hunk 4]
```diff
@@ -98,4 +100,26 @@ object io {
     implicit val outputEncoder: Encoder[Output] =
       deriveConfiguredEncoder[Output]
   }
+
+  sealed trait MetricsReporter {
+    def period: FiniteDuration
+  }
+
+  object MetricsReporter {
+    case class Stdout(period: FiniteDuration) extends MetricsReporter
```
Comment (chuwy) [inline modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/config/io.scala:109]: Could be `final`

File: modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/io/Metrics.scala [hunk 5]
```diff
@@ -64,21 +66,24 @@ object Metrics {
...
+        for {
+          logger <- Resource.liftF(Sync[F].delay(LoggerFactory.getLogger(LoggerName)))
+          reporter <- Resource.fromAutoCloseable(Sync[F].delay(Slf4jReporter.forRegistry(registry).outputTo(logger).build))
+        } yield reporter
+      case statsd: MetricsReporter.StatsD =>
+        StatsDReporter.resource(statsd, registry)
     }
 
-  def make[F[_]: Sync](reporter: Slf4jReporter, registry: MetricRegistry): Metrics[F] =
+  def make[F[_]: Sync](reporter: ScheduledReporter, registry: MetricRegistry): Metrics[F] =
```
Comment (chuwy) [inline modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/io/Metrics.scala:86]: But `ScheduledReporter` reports on its own in a background thread IIRC? So we can remove the `report` method?

**Review by chuwy — APPROVED** [review]
Looks great! Just not sure I understand the of scheduled reporter.

File: modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/io/Metrics.scala (hunk 5 above)
Comment (istreeter) [reply modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/io/Metrics.scala:86]: A `ScheduledReporter` would normally run on its own thread.  But we choose to run it on our own thread pool using the `report` method.

This is not a new change: the old `Sf4jReporter` is a subclass of `ScheduledReporter` - so we already had this setup.

What's different now.... the StatsD reporter uses a StatsD client underneath, and that client runs on its own thread.

I think this is a good overall setup - the client's blocking operations are run on a dedicated thread, but we do not waste an extra thread for our reporters which are non-blocking.

File: modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/io/StatsDReporter.scala (hunk 2 above)
Comment (istreeter) [reply modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/io/StatsDReporter.scala:2]: Fixed in a follow-up commit

File: modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/io/Metrics.scala [hunk 6]
```diff
@@ -64,21 +66,24 @@ object Metrics {
...
-      val logger = LoggerFactory.getLogger(LoggerName)
-      val reporter = Slf4jReporter.forRegistry(registry).outputTo(logger).build()
-      (reporter, registry)
+  def resource[F[_]: Sync](config: MetricsReporter): Resource[F, Metrics[F]] =
+    for {
+      registry <- Resource.liftF(Sync[F].delay((new MetricRegistry())))
+      rep <- reporter(config, registry)
+    } yield make(rep, registry)
+
+  def reporter[F[_]: Sync](config: MetricsReporter, registry: MetricRegistry): Resource[F, ScheduledReporter] =
```
Comment (benjben) [inline modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/io/Metrics.scala:75]: Maybe we could call `report` when closing the `Resource` ? Line 66 says so but I don't see it happening.

**Review by benjben — APPROVED** [review]

**Pull request closed without merging**